### ISSUE TITLE
No more websockets after upgrading to SignalR 1.1 & .net Core 2.2

### ISSUE THREAD

**Issue by StefanKoenigMUC** [aspnet/AspNetCore#4389]
Hi all,

this morning i upgraded my application to SignalR 1.1 (by this i was force to upgrade from .net core 2.1.5 to .net core 2.2).

Since those upgrades, no more Websocket connections are negotiated between my Client (Angular, also using SignalR 1.1 Libs) and my .net core server.

I've honestly no clue whats causing the problem, as there are no significant changes to the application. 

Due to the upgrades, some changes to my CORS policy where necessarcy, it looks like this right now: 

        
 
```
               services.AddCors(o => o.AddPolicy("MyPolicy", =>
              {

                    builder

                           .SetIsOriginAllowed(isOriginAllowed: _ => true)

                           .WithMethods("GET", "PUT", "POST", "DELETE", "PATCH", "OPTIONS")

                           .AllowCredentials()

                           .AllowAnyHeader();

                           .WithHeaders("Authorization", "Content-Type", "X-Requested-With",);

;                }));
```

SignalR is configured like this: 
           
```
services.AddSignalR().AddMessagePackProtocol();

         app.UseSignalR(routes =>

            { // some routes}
```

Does anyone have an idea regarding my problem?

Thanks in advance!

**Comment by pranavkm** [comment]
As of 2.2.0, the CORS middleware no longer reflects the client Origin header when credentials are allowed (See #3106). You must specify one of more Origins `AllowOriigns("https://api.originA.com", "https://www.originB.com")` as part of configuring the policy.

**Comment by StefanKoenigMUC** [comment]
Thanks for your reply, but the Allow Origins issue should already be covered by this Lambda? 

`.SetIsOriginAllowed(isOriginAllowed: _ => true)`

**Comment by pranavkm** [comment]
This is equivalent to writing `AllowAnyOrigin` - the server does not have a list of hosts that it can include in the response. Prior to 2.2, the server echoed the Origin header specified by the client which is considered dangerous

**Comment by BrennanConroy** [comment]
That looks like your server doesn't support WebSockets, what is your server environment?

**Comment by StefanKoenigMUC** [comment]
.net core (upgraded before creating this ticket from 2.1.5 to 2.2), i assume something has changed by this update, but 'ive no idea what.

websockets are enabled like that:

 ```
 public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseExceptionHandler("/Error");
                //app.UseHsts();
            }
            app.UseCors("MyPolicy");


            app.UseWebSockets();
            app.UseAuthentication();

            app.UseResponseCompression();


            app.UseHttpsRedirection();
            app.UseStaticFiles();
            app.UseCookiePolicy();

            app.UseMvc();


            app.UseSignalR(routes =>
            {
                routes.MapHub<UpdatePortfolioMessageHub>("/MessageHub");
            });
        }
```

and i'm using version 2.2 of all those Webpack packages available:

```
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.App" />
    <PackageReference Include="Microsoft.AspNetCore.SignalR" Version="1.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.SignalR.Client.Core" Version="1.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.SignalR.Common" Version="1.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.SignalR.Core" Version="1.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.SignalR.Protocols.Json" Version="1.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.SignalR.Protocols.MessagePack" Version="1.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.SignalR.Redis" Version="1.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.WebSockets" Version="2.2.0" />
    <PackageReference Include="Microsoft.AspNetCore.WebUtilities" Version="2.2.0" />
  </ItemGroup>
```

**Comment by BrennanConroy** [comment]
What OS? Are you hosting behind a proxy? IIS?

**Comment by StefanKoenigMUC** [comment]
Windows 10, No Proxy, No IIS (using the built in Webserver Kestrel).

But as you mentioned it, i tried to switch to my local IIS and now it's working again. I'll have a deeper look into it.

**Comment by StefanKoenigMUC** [comment]
Finally i was not able to find the source of the problem, switched finally my dev-env. to IIS and now everything is working as usual again.

**Comment by BrennanConroy** [comment]
Ok, to summarize, you're on Win 10 hosting with raw kestrel. You switched from 2.1.5 to 2.2 and now your SignalR connection no longer negotiates WebSockets. You then host behind IIS with the same machine/setup and are able to negotiate WebSockets on your SignalR connection?

Could you try something really quickly for me, not using IIS try to connect to the server with `skipNegotiation` set to true and WebSockets selected as the transport to connect with.
```javascript
let connection = new signalR.HubConnectionBuilder()
.withUrl("/myhub", {
    skipNegotiation: true,
    transport: signalR.HttpTransportType.WebSockets
})
.build();
```

**Comment by BrennanConroy** [comment]
cc @shirhatti

**Comment by Tratcher** [comment]
This sounds like the HTTP/2 regression we fixed in RTM. Maybe it didn't make it in?

**Comment by BrennanConroy** [comment]
Nupkg has the change so it's not that.

**Comment by blazkablatnik** [comment]
Hey !

I have the exact same problem as @StefanKoenigMUC. After upgrading to SignalR 1.1 & .net Core 2.2 (and taking care of CORS headers) negotiation fails with status "**405 Method Not Allowed**". Problem occures only on my client (using "@aspnet/signalr": "1.1.0"). It works as usual when connecting with .NET core client (using "Microsoft.AspNetCore.SignalR.Client").

When setting `skipNegotiation: true` and `transport: signalR.HttpTransportType.WebSockets` on my client, connection is successfully established. 

Is this a regression or a design change? How should I go about fixing this? Thanks :)

**Comment by BrennanConroy** [comment]
Your issue looks very different, could you please open a new issue and provide client/server logs or a repro app if possible?

**Comment by Cyrill1234** [comment]
[works.zip](https://github.com/aspnet/AspNetCore/files/2729627/works.zip)
[does not work.zip](https://github.com/aspnet/AspNetCore/files/2729634/does.not.work.zip)

I have the same problem. It works with V2.1 and it does not work with V2.2.
Here is a example.

**Comment by StefanKoenigMUC** [comment]
oh, sorry i have not seen your reply :-( - is that test still necessary?

**Comment by BrennanConroy** [comment]
@Cyrill1234 Your app doesn't work because CORS changed between 2.1 and 2.2, you can use `AllowAnyOrigin` with `AllowCredentials` see https://docs.microsoft.com/en-us/aspnet/core/migration/21-to-22?view=aspnetcore-2.2&tabs=visual-studio#update-cors-policy

@StefanKoenigMUC Yes that would be helpful, try it without hosting behind IIS. Also, is it possible for you to make a repro app?

**Comment by StefanKoenigMUC** [comment]
ok, might take until next week, as i'm off work at the moment - maybe i find some time in the evenings.

**Comment by Looooooka** [comment]
Same problem here. Luckily I came upon this thread and stopped worrying so much.
Debugging straight from net core(commandName: "project") it doesn't work even after adding any possible combination of urls in the WithOrigins command.
Debugging from IIS(commandName: "IIS")...it works as before.

**Comment by nuteman1** [comment]
I have the same problem (websockets not working) after upgrading the libraries mentioned above.
In my case its failing on the call from the azure app service to the azure signalr service. When I change the client to limit to long polling, all works and the client connection count in the azure signalr service is incremented by one, as expected.
**Any projected fix date?**

**Comment by nuteman1** [comment]
> ```js
>     skipNegotiation: true,
>     transport: signalR.HttpTransportType.WebSockets
> ```

I tried the suggestion above - still problematic, so I just leave out those settings and the problem remains.

The error I get is:

[2019-02-05T10:10:12.918Z] Debug: Starting HubConnection.
signalr.js:2657 [2019-02-05T10:10:12.919Z] Debug: Starting connection with transfer format 'Text'.
signalr.js:2657 [2019-02-05T10:10:12.919Z] Debug: Sending negotiation request: https://signalrappservice.azurewebsites.net/SignalR.AppService/PanelsHub/negotiate.
signalr.js:2657 [2019-02-05T10:10:15.809Z] Debug: Sending negotiation request: https://signalrpoc.service.signalr.net/client/negotiate?hub=panelshub&asrs.op=%2FSignalR.AppService%2FPanelsHub.
signalr.js:2657 [2019-02-05T10:10:17.121Z] Debug: Selecting transport 'WebSockets'.
signalr.js:2657 [2019-02-05T10:10:17.123Z] Trace: (WebSockets transport) Connecting.
signalr.js:3896 WebSocket connection to 'wss://signalrpoc.service.signalr.net/client/?hub=panelshub&asrs.op=%2FSignalR.AppService%2FPanelsHub&id=4ZzI6d03EKgnu-zfNMGaGw&access_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhc3JzLnMuYXV0IjoiQXV0aGVudGljYXRpb25UeXBlcy5GZWRlcmF0aW9uIiwiaHR0cDovL3NjaGVtYXMueG1sc29hcC5vcmcvd3MvMjAwNS8wNS9pZGVudGl0eS9jbGFpbXMvYXV0aGVudGljYXRpb24iOiJjSEpGTTJKVFZYWlRPVTl4UkVvekx6RnhZME40Y1dSVFNVOVdLMWRIWnk5Nk1HSk1jeXN6VWprME9EMDZibUYwYUdGdVlXSkFaWE56Wlc1alpTMW5jbkF1WTI5dCIsImV4cCI6MTU0OTM2NDk5NSwiaXNzIjoiV2VSU2lnbmFsUlNlcnZlciIsImF1ZCI6WyJXZVJTaWduYWxSQ2xpZW50cyIsImh0dHBzOi8vc2lnbmFscnBvYy5zZXJ2aWNlLnNpZ25hbHIubmV0L2NsaWVudC8_aHViPXBhbmVsc2h1YiJdLCJhc3JzLnMuaWQiOiJsUnJCVGY4bEFBQT0iLCJuYmYiOjE1NDkzNjEzOTUsImlhdCI6MTU0OTM2MTM5NX0.mpf6ekSvhm2wE-gMFjZaGJIKPNA3BPJnaf9Av8Bbd3o' failed: **HTTP Authentication failed; no valid credentials available**

It seems that when the websocket call is passed to the signalr service, despite the fact that the access_token is passed as expected (see above), the call fails with the error above.
Again, this problem does not happen with long polling, but we consider the websockets a critical part of the solution.

_Would somebody from the team please verify if this is being worked on (or not)?_

**Comment by BrennanConroy** [comment]
@nuteman1 Can you please open a new issue describing the problem you're having. It's definitely not the same issue as OP.

**Comment by BrennanConroy** [comment]
@StefanKoenigMUC We haven't been able to repro this and haven't heard from you in a while. I'll close this issue tomorrow if there is no response.

**Comment by StefanKoenigMUC** [comment]
Sorry for not giving feedback for so long. Some troubles within the last months, so that i'm was not able to continue here.

Your workaround changes the error to something else:

```
2019-02-21T14:01:25.875Z ERROR [main.js:1044] Error: Error parsing handshake response: TypeError: Right-hand side of 'instanceof' is not callable
    at HubConnection.push../node_modules/@aspnet/signalr/dist/esm/HubConnection.js.HubConnection.processHandshakeResponse (HubConnection.js:368)
    at HubConnection.push../node_modules/@aspnet/signalr/dist/esm/HubConnection.js.HubConnection.processIncomingData (HubConnection.js:318)
    at WebSocketTransport.HubConnection.connection.onreceive (HubConnection.js:65)
    at WebSocket.webSocket.onmessage [as __zone_symbol__ON_PROPERTYmessage] (WebSocketTransport.js:107)
    at WebSocket.wrapFn (zone.js:1332)
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:423)
    at Object.onInvokeTask (core.js:17280)
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:422)
    at Zone.push../node_modules/zone.js/dist/zone.js.Zone.runTask (zone.js:195)
    at ZoneTask.push../node_modules/zone.js/dist/zone.js.ZoneTask.invokeTask [as invoke] (zone.js:498).
```
Currently used versions are: 

```
    "@aspnet/signalr": "^1.1.2",
    "@aspnet/signalr-protocol-msgpack": "^1.1.0",
```

using Windows 2016 with IIS 10. As there should now be more time, i will prepare an example during the next days.

**Comment by BrennanConroy** [comment]
What kind of environment are you running the client in?

It looks like either `data instanceof ArrayBuffer` or `data instanceof Buffer` is failing, what do those types look like in your environment?

**Comment by IVL-Mukesh** [comment]
We are also experience same issue that mention by @StefanKoenigMUC . Please help us to fix this...Thanks

**Comment by jtbrower** [comment]
Regarding the "Error parsing handshake response: TypeError: Right-hand side of 'instanceof' is not callable" error I thought I would share with everyone what was causing this error for me, though I realize that some details of my problem are different.

I am serving up both MessagePack and JSON capable endpoints using NetCore 3.Preview+ (debugged through Visual Studio 2019 Preview) while my Angular 7 client SPA is using signalR 1.1.2 / signalr-protocol-msgpack 1.1.0 via Chrome 72.0.3626.121.

When I first enabled MessagePack on the client I had to introduce a work-around that @BrennanConroy and @andreatosato provided for issue #2736 where a script block was added to the index.html and polyfills.ts files of Angular.  Although that work-around solved my MessagePack error,  I found that if I switch the client back to use JSON instead of MessagePack I had to remove the script block from index.html or the "Right-hand side of 'instanceof' is not callable" exception would be thrown.  

To summarize, the following script block in index.html will cause the error when using JSON but is required when using MessagePack.  The pollyfill.ts modification can remain for both MessagePack and JSON enabled clients.

**Comment by StefanKoenigMUC** [comment]
i finally figured it out, that error occurs for me: error occured only, when MessagePack is just installed on ServerSide of the application.
Adding it to the client, makes it work again.

**Comment by anurse** [comment]
Yeah, the MessagePack library brings along polyfills for `ArrayBuffer` and `Buffer`. It sounds like we can make a small improvement here by checking for the existance of these types before running `instanceof` in https://github.com/aspnet/AspNetCore/blob/321327f84b2b22dcff2e9beb06a9a64236c5cced/src/SignalR/clients/ts/signalr-protocol-msgpack/src/Utils.ts#L6 .

We'll look improving the error messaging for this in a future release

**Comment by anurse** [comment]
> `var Buffer = Buffer || [];`

It looks like this polyfill is what causes the problem. We check `Buffer` against `undefined` but if the polyfill is triggered, `Buffer` is set to `[]` which means `typeof Buffer` returns `object`. That means we get past the `typeof Buffer` check and then try to use it with `instanceof`, which fails.

I think we need a new approach to this one. Is someone able to provide a clonable sample that reproduces the problem?

A longer term plan may be to consider a new MessagePack library that doesn't depend on `Buffer` at all, such as https://github.com/msgpack/msgpack-javascript

**Comment by ArntB** [comment]
If you set up a angular cli project and add signalr to it you will get this error becaus it adds the folowing script to index.html
```
<script>
    var global = global || window;
    var Buffer = Buffer || [];
    var process = process || {
      env: { DEBUG: undefined },
      version: []
    };
  </script>
```

**Comment by anurse** [comment]
Closing as this appears to be an issue with the polyfill that was provided in a comment above. If we can get more context on this polyfill as requested above we can see if there's something we can do to fix it.
